### PR TITLE
improve detection of PGP messages

### LINKED ISSUE #715: improve detection of encrypted messages
Please go through compatibility account and make sure that most encrypted messages get a "encrypted shield badge" in the inbox.

Therefore if user filters for encrypted messages only, they will show up.

There are probably many message formats that are not recognized today.

### PR DESCRIPTION
**Pull request by DenBond7**
This PR improved the detection of PGP messages

close #715 

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities

### CODE REVIEW

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/email/gmail/GmailApiHelper.kt [hunk 1]
```diff
@@ -86,8 +86,10 @@ import javax.net.ssl.SSLException
 class GmailApiHelper {
   companion object {
     const val DEFAULT_USER_ID = "me"
-    const val PATTERN_SEARCH_ENCRYPTED_MESSAGES =
-      "PGP OR GPG OR OpenPGP OR filename:asc OR filename:message OR filename:pgp OR filename:gpg"
+    const val PATTERN_SEARCH_PGP =
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/email/gmail/GmailApiHelper.kt:89]: @sosnovsky Please pay attention on this one

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/com/google/api/services/gmail/model/MessageExt.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package com.flowcrypt.email.extensions.com.google.api.services.gmail.model
+
+import com.flowcrypt.email.api.email.EmailUtil
+import com.flowcrypt.email.extensions.kotlin.asContentTypeOrNull
+import com.google.api.services.gmail.model.Message
+
+/**
+ * @author Denys Bondarenko
+ */
+fun Message.hasPgp(): Boolean {
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/com/google/api/services/gmail/model/MessageExt.kt:15]: @sosnovsky Please pay attention on this one

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/email/EmailUtil.kt [hunk 3]
```diff
@@ -614,19 +615,35 @@ class EmailUtil {
...
-     * @return A generated [SearchTerm].
-     */
-    fun genEncryptedMsgsSearchTerm(account: AccountEntity): SearchTerm {
+    fun hasSignedData(rawMsg: String?) =
+      rawMsg?.contains("-----BEGIN PGP SIGNED MESSAGE-----") == true
+
+    fun genPgpThingsSearchTerm(account: AccountEntity): SearchTerm {
       return if (AccountEntity.ACCOUNT_TYPE_GOOGLE.equals(account.accountType, ignoreCase = true)) {
-        GmailRawSearchTerm(GmailApiHelper.PATTERN_SEARCH_ENCRYPTED_MESSAGES)
+        GmailRawSearchTerm(GmailApiHelper.PATTERN_SEARCH_PGP)
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/email/EmailUtil.kt:626]: @sosnovsky Please pay attention to this one

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/jakarta/mail/PartExt.kt [hunk 4]
```diff
@@ -89,3 +90,29 @@ fun Part.hasPgpThings(): Boolean {
   val detectedBlocks = RawBlockParser.detectBlocks(this)
   return detectedBlocks.any { it.type in RawBlockParser.PGP_BLOCK_TYPES }
 }
+
+fun Part.hasPgp(): Boolean {
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/jakarta/mail/PartExt.kt:94]: @sosnovsky Please pay attention to this one

**Comment by DenBond7** [comment]
@sosnovsky I've improved PGP identification. Also, I've marked important places for you.

In general, to identify PGP things the app uses the following logic:

- Analyze the message structure to identify `PGP/MIME`(encrypted or signed)
- Analyze the message attachments(based on extensions)
- Analyze the message body text(find `-----BEGIN PGP MESSAGE-----` or `-----BEGIN PGP SIGNED MESSAGE-----`)

It works a bit differently for `Gmail API` and `IMAP`

`Gmail API`

Here the app doesn't make additional requests. It's fast but mayn't provide 100% matching. Because we use `message.snippet` to check for `-----BEGIN PGP MESSAGE-----` or `-----BEGIN PGP SIGNED MESSAGE-----`.

`IMAP`

Here the app after fetching the base message details also makes an additional request(`IMAP SEARCH`) to check the message body(whole message) and find `-----BEGIN PGP MESSAGE-----` or `-----BEGIN PGP SIGNED MESSAGE-----`. 

Please try the added changes while I'm working on fixing some tests.

**Comment by sosnovsky** [comment]
Hi @DenBond7, I tested current implementation and it works very well - I haven't noticed any issues with incorrect `encrypted/non-encrypted` message badges, previously mentioned messages are correctly rendered as `encrypted` now 👍

I've just noticed non-related issue - user image is rendered as empty circle now (as I remember, previously it showed 2 first letters of user name, like in the messages list):

Screenshot 2024-05-27 at 11 51 45

**Comment by DenBond7** [comment]
@sosnovsky Please look at this one when you will have some free time. It's blocking me a little ))

**Review by sosnovsky — APPROVED** [review]
It worked well when I've checked it before, and as your latest changes affected mostly tests, I think it's ok to merge this PR to stop blocking you.
I'll re-test later today in `master`, and will let you know if will notice something.

**Pull request merged**